File: solutions/python/vendingmachine/ready_state.py
from vending_machine_state import VendingMachineState

class ReadyState(VendingMachineState):
    def select_product(self, product):
        print("Product already selected. Please make payment.")

    def insert_coin(self, coin):
        self.vending_machine.add_coin(coin)
        print(f"Coin inserted: {coin}")
        self.check_payment_status()

    def insert_note(self, note):
        self.vending_machine.add_note(note)
        print(f"Note inserted: {note}")
        self.check_payment_status()

    def dispense_product(self):
        print("Please make payment first.")

    def return_change(self):
        change = self.vending_machine.total_payment - self.vending_machine.selected_product.price
        if change > 0:
            print(f"Change returned: ${change}")
            self.vending_machine.reset_payment()
        else:
            print("No change to return.")
        self.vending_machine.state = self.vending_machine.idle_state

    def check_payment_status(self):
        if self.vending_machine.total_payment >= self.vending_machine.selected_product.price:
            self.vending_machine.state = self.vending_machine.dispense_state

File: solutions/csharp/vendingmachine/ReadyState.cs
using System;

namespace VendingMachineApp
{
    public class ReadyState : IVendingMachineState
    {
        private readonly VendingMachine _vendingMachine;

        public ReadyState(VendingMachine vendingMachine)
        {
            _vendingMachine = vendingMachine;
        }

        public void SelectProduct(Product product)
        {
            Console.WriteLine("Product already selected. Please make payment.");
        }

        public void InsertCoin(Coin coin)
        {
            _vendingMachine.AddCoin(coin);
            Console.WriteLine("Coin inserted: " + coin);
            CheckPaymentStatus();
        }

        public void InsertNote(Note note)
        {
            _vendingMachine.AddNote(note);
            Console.WriteLine("Note inserted: " + note);
            CheckPaymentStatus();
        }

        public void DispenseProduct()
        {
            Console.WriteLine("Please make payment first.");
        }

        public void ReturnChange()
        {
            double change = _vendingMachine.TotalPayment - _vendingMachine.SelectedProduct.Price;
            if (change > 0)
            {
                Console.WriteLine("Change returned: $" + change);
                _vendingMachine.ResetPayment();
            }
            else
            {
                Console.WriteLine("No change to return.");
            }
            _vendingMachine.SetState(_vendingMachine.GetIdleState());
        }

        private void CheckPaymentStatus()
        {
            if (_vendingMachine.TotalPayment >= _vendingMachine.SelectedProduct.Price)
            {
                _vendingMachine.SetState(_vendingMachine.GetDispenseState());
            }
        }
    }
}